pss/pss.go

```diff
@@ -553,20 +555,58 @@ func (p *Pss) isSelfPossibleRecipient(msg *PssMsg, prox bool) bool {
 // SECTION: Message sending
 /////////////////////////////////////////////////////////////////////
 
-func (p *Pss) enqueue(msg *PssMsg) error {
+// Enqueue a message msg. Pending will be true for messages with a booked slot in the queue
+func (p *Pss) enqueue(msg *PssMsg, pending bool) error {
 	defer metrics.GetOrRegisterResettingTimer("pss.enqueue", nil).UpdateSince(time.Now())
 
 	outboxmsg := newOutboxMsg(msg)
-	select {
-	case p.outbox <- outboxmsg:
-		metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
-
-		return nil
-	default:
+	// We mutex between the calculation of the outbox capacity and the introduction in the channel so we make sure nobody
+	// add a message to the queue after checking outboxCapacity
+	mux := sync.Mutex{}
+	mux.Lock()
+	defer mux.Unlock()
+	pendingSize := p.getPending()
+	// Only allow defaultOutboxCapacity messages at most processed (both enqueued or being forwarded)
+	if pending || pendingSize < defaultOutboxCapacity { //If pending there is already a slot booked for this message
+		if !pending {
+			// We book a slot in the queue increasing pending messages and release it after successfully sending the message
+			p.increasePending()
+		}
+		select {
+		case p.outbox <- outboxmsg:
+			metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
+			return nil
+		default:
+		}
 	}
-
 	metrics.GetOrRegisterCounter("pss.enqueue.outbox.full", nil).Inc(1)
-	return errors.New("outbox full")
+	if pending {
+		return errors.New("unexpected outbox full for pending message!")
+	} else {
+		return errors.New("outbox full")
+	}
+}
+
+// Access to forwardPending is protected with RWMutex
+
+func (p *Pss) increasePending() {
+	p.pendingMutex.Lock()
+	defer p.pendingMutex.Unlock()
+	metrics.GetOrRegisterCounter("pss.forward.pending", nil).Inc(1)
+	p.forwardPending++
+}
+
+func (p *Pss) decreasePending() {
+	p.pendingMutex.Lock()
+	defer p.pendingMutex.Unlock()
+	metrics.GetOrRegisterCounter("pss.forward.pending", nil).Dec(1)
+	p.forwardPending--
+}
+
+func (p *Pss) getPending() int {
+	p.pendingMutex.Lock()
```

Comment:
`p.pendingMutex.RLock()` ?

Reply:
Fixed in commit 462897d
Also added unit tests for enqueue function.